File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> v;
        int pSize = p.size();
        int sSize = s.size();
        if(pSize > sSize) return v;
        vector<int> a(26, 0), b(26, 0);
        for(int i=0;i<pSize;i++){
            a[s[i]-'a']++;
            b[p[i]-'a']++;
        }
        int left = 0;
        for(int i=pSize;i<sSize;i++){
            if(a == b) v.push_back(left);
            a[s[left++] - 'a']--;
            a[s[i]- 'a']++;
        }
        if(a == b) v.push_back(left);
        return v;
    }
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int a[] = new int[26], b[] = new int[26];
        int pSize = p.length();
        for(int i=0;i<pSize;i++){
            b[p.charAt(i)-'a']++;
        }
        int start = 0;
        for(int i=0;i<s.length();i++){
            if(i >= pSize) a[s.charAt(start++)-'a']--;
            a[s.charAt(i)-'a']++;
            if(Arrays.equals(a, b)) res.add(start);
        }
        return res;
    }
}